### PR TITLE
Fixed boat not taking care of block-liquid hooks

### PR DESCRIPTION
**Pull request by CreativeMD**
Hey there,
this pr is a follow up of (https://github.com/MinecraftForge/MinecraftForge/pull/4604). Basically the issue is that a boat does not take care of the new added liquid detection. This pr changes that.

Unfortunately there is no way to get the liquid height, so I had to add another method for that in `Block`:
```java
/**
* Called when entities try to swim (for example a boat in water)
* 
* @param world world that is being tested.
* @param pos block thats being tested.
* @param state state at world/pos
* @param material liquid thats being tested.
* @return relative height of the given liquid (material), a value between 0 and 1
*/
public float getBlockLiquidHeight(World world, BlockPos pos, IBlockState state, Material material)
{
     return 0;
}
```

Everything else is pretty straight forward and follows the same way as the previous pr.

In Regards
CreativeMD

### CODE REVIEW

**Comment by CreativeMD** [comment]
In theory, the pr is ready.

**Comment by bs2609** [comment]
This doesn't address Forge fluid blocks.

**Comment by CreativeMD** [comment]
I thought Forge fluid blocks extend BlockLiquid?

**Comment by bs2609** [comment]
Pretty clear they don't:
https://github.com/MinecraftForge/MinecraftForge/blob/857573f10fbb18b3b15d7053757a2216730b459d/src/main/java/net/minecraftforge/fluids/BlockFluidBase.java#L66

**Comment by CreativeMD** [comment]
Oh right, thanks for letting me know. I added it now.

File: src/main/java/net/minecraftforge/fluids/BlockFluidBase.java [hunk 1]
```diff
@@ -802,4 +802,10 @@ private boolean isWithinFluid(IBlockAccess world, BlockPos pos, Vec3d vec)
         return filled < 0 ? vec.y > pos.getY() + filled + 1
                           : vec.y < pos.getY() + filled;
     }
+    
+    @Override
+    public float getBlockLiquidHeight(World world, BlockPos pos, IBlockState state, Material material)
+    {
+        return getFilledPercentage(world, pos);
```
Comment (bs2609) [inline src/main/java/net/minecraftforge/fluids/BlockFluidBase.java]: What are you planning to do here for negative-density fluids?

Specifically:
https://github.com/MinecraftForge/MinecraftForge/blob/857573f10fbb18b3b15d7053757a2216730b459d/src/main/java/net/minecraftforge/fluids/IFluidBlock.java#L79-L80

File: patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch [hunk 2]
```diff
@@ -9,7 +22,42 @@
...
+                             float f = BlockLiquid.func_190972_g(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos);
+@@ -594,6 +609,18 @@
+                         blockpos$pooledmutableblockpos.func_181079_c(k1, l1, i2);
+                         IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
+ 
++                        Boolean result = iblockstate.func_177230_c().isAABBInsideMaterial(field_70170_p, blockpos$pooledmutableblockpos, axisalignedbb, Material.field_151586_h);
++                        if (result != null) {
++                            if (!result) continue;
++                            
++                            if(iblockstate.func_177230_c().getBlockLiquidHeight(field_70170_p, blockpos$pooledmutableblockpos, iblockstate, Material.field_151586_h) + blockpos$pooledmutableblockpos.func_177956_o() > 0)
```
Comment (bs2609) [inline patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch]: This doesn't seem correct

File: src/main/java/net/minecraftforge/fluids/BlockFluidBase.java (hunk 1 above)
Comment (CreativeMD) [reply src/main/java/net/minecraftforge/fluids/BlockFluidBase.java]: Oh right, maybe return zero?

**Comment by CreativeMD** [comment]
@bs2609 so is it ready now?

**Comment by CreativeMD** [comment]
Fixed all merge conflicts.

File: src/main/java/net/minecraftforge/fluids/BlockFluidBase.java [hunk 3]
```diff
@@ -807,4 +807,11 @@ private boolean isWithinFluid(IBlockAccess world, BlockPos pos, Vec3d vec)
         return filled < 0 ? vec.y > pos.getY() + filled + 1
                           : vec.y < pos.getY() + filled;
     }
+    
+    @Override
+    public float getBlockLiquidHeight(World world, BlockPos pos, IBlockState state, Material material)
+    {
+        float filled = getFilledPercentage(world, pos);
+        return filled > 0 ? filled : 0;
```
Comment (tterrag1098) [inline src/main/java/net/minecraftforge/fluids/BlockFluidBase.java]: Use `Math.max`

File: patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch [hunk 4]
```diff
@@ -9,7 +22,42 @@
...
++
+                         if (iblockstate.func_185904_a() == Material.field_151586_h)
+                         {
+                             float f = BlockLiquid.func_190972_g(iblockstate, this.field_70170_p, blockpos$pooledmutableblockpos);
+@@ -594,6 +609,18 @@
+                         blockpos$pooledmutableblockpos.func_181079_c(k1, l1, i2);
+                         IBlockState iblockstate = this.field_70170_p.func_180495_p(blockpos$pooledmutableblockpos);
+ 
++                        Boolean result = iblockstate.func_177230_c().isAABBInsideMaterial(field_70170_p, blockpos$pooledmutableblockpos, axisalignedbb, Material.field_151586_h);
++                        if (result != null) {
```
Comment (tterrag1098) [inline patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch:46]: This pattern is repeated here 3 times and presumably elsewhere. I hate the "`Boolean` as trinary" pattern, but I suppose we're stuck with it until 1.13. Could this logic be deduped somehow?

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 5]
```diff
@@ -1403,6 +1403,20 @@
 +     {
 +         return null;
 +     }
++     
++     /**
++      * Called when entities try to swim (for example a boat in water)
```
Comment (tterrag1098) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: This documentation seems incomplete.

**Review by tterrag1098 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch (hunk 4 above)
Comment (CreativeMD) [reply patches/minecraft/net/minecraft/entity/item/EntityBoat.java.patch:46]: I don't know, I just tried to continue this way. Those methods are not added by this pr, so I'm not sure if this is the right place to change them.

**Review by tterrag1098 — APPROVED** [review]

**Review by mezz — APPROVED** [review]

**Pull request merged**